File: libs/moqtail-ts/src/model/data/subgroup_header.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ByteBuffer, FrozenByteBuffer, BaseByteBuffer } from '../common/byte_buffer'
import { ProtocolViolationError } from '../error'
import { SubgroupHeaderType } from './constant'

// TODO: couple type and subgroup id
export class SubgroupHeader {
  readonly subgroupId: bigint | undefined
  readonly trackAlias: bigint
  readonly groupId: bigint

  constructor(
    readonly type: SubgroupHeaderType,
    trackAlias: bigint | number,
    groupId: bigint | number,
    subgroupId: bigint | number | undefined,
    readonly publisherPriority: number,
  ) {
    this.trackAlias = BigInt(trackAlias)
    this.groupId = BigInt(groupId)
    if (subgroupId !== undefined) {
      this.subgroupId = BigInt(subgroupId)
    } else {
      this.subgroupId = subgroupId
    }
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(this.type)
    buf.putVI(this.trackAlias)
    buf.putVI(this.groupId)
    if (SubgroupHeaderType.hasExplicitSubgroupId(this.type)) {
      if (this.subgroupId === undefined) {
        throw new ProtocolViolationError(
          'SubgroupHeader.serialize',
          'Subgroup_id field is required for this header type',
        )
      }
      buf.putVI(this.subgroupId)
    }
    buf.putU8(this.publisherPriority)
    return buf.freeze()
  }

  static deserialize(buf: BaseByteBuffer): SubgroupHeader {
    const headerType = SubgroupHeaderType.tryFrom(buf.getNumberVI())
    const trackAlias = buf.getVI()
    const groupId = buf.getVI()
    let subgroupId: bigint | undefined
    if (SubgroupHeaderType.hasExplicitSubgroupId(headerType)) {
      subgroupId = buf.getVI()
    } else if (headerType === SubgroupHeaderType.Type0x08 || headerType === SubgroupHeaderType.Type0x09) {
      subgroupId = 0n
    }
    const publisherPriority = buf.getU8()
    return new SubgroupHeader(headerType, trackAlias, groupId, subgroupId, publisherPriority)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('SubgroupHeader', () => {
    test('roundtrip serialization/deserialization', () => {
      const headerType = SubgroupHeaderType.Type0x0C
      const trackAlias = 87n
      const groupId = 9n
      const subgroupId = 11n
      const publisherPriority = 255
      const header = new SubgroupHeader(headerType, trackAlias, groupId, subgroupId, publisherPriority)
      const frozen = header.serialize()
      const parsed = SubgroupHeader.deserialize(frozen)
      expect(parsed.type).toBe(header.type)
      expect(parsed.trackAlias).toBe(header.trackAlias)
      expect(parsed.groupId).toBe(header.groupId)
      expect(parsed.subgroupId).toBe(header.subgroupId)
      expect(parsed.publisherPriority).toBe(header.publisherPriority)
      expect(frozen.remaining).toBe(0)
    })
    test('excess roundtrip', () => {
      const headerType = SubgroupHeaderType.Type0x0C
      const trackAlias = 87n
      const groupId = 9n
      const subgroupId = 11n
      const publisherPriority = 255
      const header = new SubgroupHeader(headerType, trackAlias, groupId, subgroupId, publisherPriority)
      const serialized = header.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const parsed = SubgroupHeader.deserialize(frozen)
      expect(parsed.type).toBe(header.type)
      expect(parsed.trackAlias).toBe(header.trackAlias)
      expect(parsed.groupId).toBe(header.groupId)
      expect(parsed.subgroupId).toBe(header.subgroupId)
      expect(parsed.publisherPriority).toBe(header.publisherPriority)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })
    test('partial message fails', () => {
      const headerType = SubgroupHeaderType.Type0x0C
      const trackAlias = 87n
      const groupId = 9n
      const subgroupId = 11n
      const publisherPriority = 255
      const header = new SubgroupHeader(headerType, trackAlias, groupId, subgroupId, publisherPriority)
      const serialized = header.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        SubgroupHeader.deserialize(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/data/subgroup_header.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{Buf, BufMut, Bytes, BytesMut};

use super::constant::SubgroupHeaderType;

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct SubgroupHeader {
  pub header_type: SubgroupHeaderType,
  pub track_alias: u64,
  pub group_id: u64,
  pub subgroup_id: Option<u64>,
  pub publisher_priority: u8,
}

impl SubgroupHeader {
  /// Create a new subgroup header with fixed Subgroup ID = 0
  pub fn new_fixed_zero_id(
    track_alias: u64,
    group_id: u64,
    publisher_priority: u8,
    has_extensions: bool,
  ) -> Self {
    let header_type = if has_extensions {
      SubgroupHeaderType::Type0x09
    } else {
      SubgroupHeaderType::Type0x08
    };

    Self {
      header_type,
      track_alias,
      group_id,
      subgroup_id: Some(0),
      publisher_priority,
    }
  }

  /// Create a new subgroup header where Subgroup ID = First Object ID
  pub fn new_first_object_id(
    track_alias: u64,
    group_id: u64,
    publisher_priority: u8,
    has_extensions: bool,
  ) -> Self {
    let header_type = if has_extensions {
      SubgroupHeaderType::Type0x0B
    } else {
      SubgroupHeaderType::Type0x0A
    };

    Self {
      header_type,
      track_alias,
      group_id,
      subgroup_id: None, // Will be set to first object ID
      publisher_priority,
    }
  }

  /// Create a new subgroup header with explicit Subgroup ID
  pub fn new_with_explicit_id(
    track_alias: u64,
    group_id: u64,
    subgroup_id: u64,
    publisher_priority: u8,
    has_extensions: bool,
  ) -> Self {
    let header_type = if has_extensions {
      SubgroupHeaderType::Type0x0D
    } else {
      SubgroupHeaderType::Type0x0C
    };

    Self {
      header_type,
      track_alias,
      group_id,
      subgroup_id: Some(subgroup_id),
      publisher_priority,
    }
  }

  pub fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();

    // Type field
    buf.put_vi(self.header_type as u64)?;

    // Track Alias
    buf.put_vi(self.track_alias)?;

    // Group ID
    buf.put_vi(self.group_id)?;

    // Subgroup ID (if present)
    if self.header_type.has_explicit_subgroup_id() {
      if let Some(id) = self.subgroup_id {
        buf.put_vi(id)?;
      } else {
        return Err(ParseError::ProtocolViolation {
          context: "SubgroupHeader::serialize(header_type)",
          details: "Subgroup_id field is required for this header type".to_string(),
        });
      }
    }

    // Publisher Priority
    buf.put_u8(self.publisher_priority);

    Ok(buf.freeze())
  }

  pub fn deserialize(bytes: &mut Bytes) -> Result<Self, ParseError> {
    // Parse type
    let type_value = bytes.get_vi()?;

    let header_type = SubgroupHeaderType::try_from(type_value)?;

    // Parse track alias
    let track_alias = bytes.get_vi()?;

    // Parse group ID
    let group_id = bytes.get_vi()?;

    // Parse subgroup ID if present
    let subgroup_id = if header_type.has_explicit_subgroup_id() {
      Some(bytes.get_vi()?)
    } else if matches!(
      header_type,
      SubgroupHeaderType::Type0x08 | SubgroupHeaderType::Type0x09
    ) {
      Some(0) // Fixed at 0 for types 0x08 and 0x09
    } else {
      None // For types 0x0A and 0x0B, subgroup ID = first object ID
    };

    if bytes.remaining() < 1 {
      return Err(ParseError::NotEnoughBytes {
        context: "SubgroupHeader::deserialize(publisher_priority)",
        needed: 1,
        available: bytes.remaining(),
      });
    }
    let publisher_priority = bytes.get_u8();

    Ok(Self {
      header_type,
      track_alias,
      group_id,
      subgroup_id,
      publisher_priority,
    })
  }
}

#[cfg(test)]
mod tests {

  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let header_type = SubgroupHeaderType::Type0x0C;
    let track_alias = 87;
    let group_id = 9;
    let subgroup_id = Some(11);
    let publisher_priority = 255;
    let subgroup_header = SubgroupHeader {
      header_type,
      track_alias,
      group_id,
      subgroup_id,
      publisher_priority,
    };

    let mut buf = subgroup_header.serialize().unwrap();
    let deserialized = SubgroupHeader::deserialize(&mut buf).unwrap();
    assert_eq!(deserialized, subgroup_header);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let header_type = SubgroupHeaderType::Type0x0C;
    let track_alias = 87;
    let group_id = 9;
    let subgroup_id = Some(11);
    let publisher_priority = 255;
    let subgroup_header = SubgroupHeader {
      header_type,
      track_alias,
      group_id,
      subgroup_id,
      publisher_priority,
    };

    let serialized = subgroup_header.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let deserialized = SubgroupHeader::deserialize(&mut buf).unwrap();
    assert_eq!(deserialized, subgroup_header);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let header_type = SubgroupHeaderType::Type0x0C;
    let track_alias = 87;
    let group_id = 9;
    let subgroup_id = Some(11);
    let publisher_priority = 255;
    let subgroup_header = SubgroupHeader {
      header_type,
      track_alias,
      group_id,
      subgroup_id,
      publisher_priority,
    };
    let buf = subgroup_header.serialize().unwrap();
    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = SubgroupHeader::deserialize(&mut partial);
    assert!(deserialized.is_err());
  }
}
